File: projects/codeeditor/CodeEditor.py
#SUCCESS
import pygame
import subprocess
class CodeEditor:
    def __init__(self,_rect,_text):
        self.rect=_rect
        self.SetText(_text)
        self.cursor=0
    def SetText(self,_text):
        self.text=_text
        self.lines=self.text.split('\n')
    def Insert(self,value):
        self.SetText(self.text[:self.cursor]+value+self.text[self.cursor:])
        self.cursor+=len(value)
    def Backspace(self):
        if self.cursor>0:
            self.SetText(self.text[:(self.cursor-1)]+self.text[self.cursor:])
            self.cursor-=1
    def CursorLeft(self):
        if self.cursor>0:
            self.cursor-=1
    def CursorRight(self):
        if self.cursor<len(self.text):
            self.cursor+=1
    def Save(self):
        File.WriteFile("projects/test/Test.cs",self.text)
    def OnEvent(self,graphics,event):
        if event.type==pygame.KEYDOWN:
            keys=pygame.key.get_pressed()
            if keys[pygame.K_LCTRL]:
                if event.key==pygame.K_TAB:
                    graphics.NextUI()
            else:
                if event.key==pygame.K_RETURN:
                    self.Insert("\n")
                elif event.key==pygame.K_TAB:
                    self.Insert("    ")
                elif event.key==pygame.K_BACKSPACE:
                    self.Backspace()
                elif event.key==pygame.K_LEFT:
                    self.CursorLeft()
                elif event.key==pygame.K_RIGHT:
                    self.CursorRight()
                else:
                    self.Insert(event.unicode)
    def Draw(self,graphics,active):
        y=self.rect[1]
        x=self.rect[0]
        bgcolor=(180,180,180) if active else (200,200,200)
        graphics.DrawRect(self.rect,bgcolor)
        textColor=(0,20,45)
        for l in self.lines:
            graphics.DrawText((x,y),l,textColor)
            y+=graphics.linesize
        if active:
            cursorYOffset=graphics.fontsize*0.1
            y=self.rect[1]
            numCharacters=0
            if len(self.lines)==0:
                graphics.DrawRect((x,y,2,graphics.fontsize),textColor)
            else:
                for l in self.lines:
                    charsInLine=len(l)
                    if self.cursor>=numCharacters and self.cursor<=numCharacters+charsInLine:
                        cursorIsCharsFromStartOfLine=self.cursor-numCharacters
                        cursorX=graphics.font.size(l[0:cursorIsCharsFromStartOfLine])[0]+x
                        graphics.DrawRect((cursorX,y+cursorYOffset,2,graphics.fontsize),textColor)
                    numCharacters+=charsInLine+1
                    y+=graphics.linesize
class GameObject:
    def __init__(self,_name):
        self.name=_name
        self.components=[]
class Button:
    def __init__(self,_rect,_text,_onclick):
        self.rect=_rect
        self.text=_text
        self.onclick=_onclick
    def Save(self):
        pass
    def OnEvent(self,graphics,event):
        if event.type==pygame.KEYDOWN:
            if event.key==pygame.K_RETURN:
                self.onclick()
            elif event.key==pygame.K_TAB:
                graphics.NextUI()
    def Draw(self,graphics,active):
        bgcolor=(180,180,180) if active else (200,200,200)
        graphics.DrawRect(self.rect,bgcolor)
        textColor=(0,50,100)
        graphics.DrawText((self.rect[0],self.rect[1]),self.text,textColor)
class TreeView:
    def __init__(self,_rect):
        self.gameObjects=[]
        self.rect=_rect
    def Save(self):
        pass
    def Add(self,gameObject):
        self.gameObjects.append(gameObject)
    def OnEvent(self,graphics,event):
        if event.type==pygame.KEYDOWN:
            if event.key==pygame.K_TAB:
                graphics.NextUI()
    def Draw(self,graphics,active):
        bgcolor=(180,180,180) if active else (200,200,200)
        graphics.DrawRect(self.rect,bgcolor)
        graphics.DrawText((self.rect[0],self.rect[1]),"TreeView",(0,50,100))
class File:
    def ReadFile(path):
        f=open(path,'r')
        content=f.read()
        f.close()
        return content
    def WriteFile(path,content):
        f=open(path,'w')
        f.write(content)
        f.close()
class Graphics:
    def __init__(self,_screenSize,_font,_fontsize):
        self.surface=pygame.display.set_mode((_screenSize[0],_screenSize[1]))
        self.font=pygame.font.Font(_font,_fontsize)
        self.fontsize=_fontsize
        self.linesize=self.fontsize*1.4
        self.screenSize=_screenSize
        self.uis=[]
        self.objects=[]
        self.activeUI=0
    def Clear(self,color):
        self.surface.fill(color)
    def Save(self):
        for u in self.uis:
            u.Save()
    def DrawText(self,position,text,color):
        img=self.font.render(text,True,color)
        self.surface.blit(img,position)
    def DrawRect(self,rect,color):
        pygame.draw.rect(self.surface,color,pygame.Rect(rect[0],rect[1],rect[2],rect[3]))
    def AddObject(self,obj):
        self.objects.append(obj)
    def DrawObjects(self):
        for obj in self.objects:
            obj.Draw()
    def ClearObjects(self):
        self.objects.clear()
    def AddUI(self,ui):
        self.uis.append(ui)
    def OnEvent(self,event):
        self.uis[self.activeUI].OnEvent(self,event)
    def DrawUI(self):
        for i in range(0,len(self.uis)):
            self.uis[i].Draw(self,i==self.activeUI)
    def NextUI(self):
        self.activeUI+=1
        if self.activeUI>=len(self.uis):
            self.activeUI=0
class Program:
    def Save(self):
        self.graphics.Save()
    def Run(self):
        self.graphics.Save()
        result=subprocess.run(["compiler\\bin\\Debug\\net8.0\\CSharpToPython.exe","projects/test/Test.cs--projects/test/Test.py"],shell=True,capture_output=True,text=True).stdout
        if result=="ERROR":
            print(File.ReadFile("projects/test/Test.py"))
        else:
            globals={"graphics":self.graphics}
            exec(File.ReadFile("projects/test/Test.py"),globals)
    def Start(self):
        pygame.init()
        self.graphics=Graphics((1280,780),"RedditMono-Medium.ttf",16)
        self.graphics.AddUI(TreeView((20,80,280,680)))
        self.graphics.AddUI(CodeEditor((320,80,900,680),File.ReadFile("projects/test/Test.cs")))
        self.graphics.AddUI(Button((500,20,100,40),">",self.Run))
        self.graphics.AddUI(Button((780,20,100,40),"save",self.Save))
        running=True
        pygame.key.set_repeat(500,25)
        while running:
            for event in pygame.event.get():
                if event.type==pygame.KEYDOWN and event.key==pygame.K_ESCAPE:
                    self.graphics.ClearObjects()
                elif event.type==pygame.QUIT:
                    running=False
                else:
                    self.graphics.OnEvent(event)
            self.graphics.Clear((220,220,220))
            self.graphics.DrawUI()
            self.graphics.DrawObjects()
            pygame.display.flip()
        pygame.quit()
    def Main():
        program=Program()
        program.Start()
Program.Main()

File: projects/codeeditor/CodeEditor.cs
using pygame;
using subprocess;

class CodeEditor {
    Rect rect;
    string text;
    int cursor;
    string[] lines;
    int scrollY;

    CodeEditor(Rect _rect, string _text){
        rect = _rect;
        SetText(_text);
        cursor = 0;
        scrollY = 0;
    }

    void SetText(string _text){
        text = _text;
        lines = text.split('\n');
    }

    void Insert(string value){
        SetText(text[..cursor] + value + text[cursor..]);
        cursor+=len(value);
    }

    void Backspace(){
        if(cursor>0){
            SetText(text[..(cursor-1)] + text[cursor..]);
            cursor--;
        }
    }

    void SetCursor(int _cursor){
        cursor = _cursor;
        var locationID = GetCursorLocationID();
        if(locationID[1]>scrollY+20){
            scrollY = locationID[1] - 20;
        }
        else if(locationID[1]<scrollY+10){
            scrollY = max(locationID[1] - 10, 0);
        }
    }

    void CursorLeft(){
        if(cursor>0){
            SetCursor(cursor-1);
        }
    }

    void CursorRight(){
        if(cursor<len(text)){
            SetCursor(cursor+1);
        }    
    }

    void Save(){
        File.WriteFile("projects/test/Test.cs", text);
    }

    void OnEvent(Graphics graphics, Event @event){
        if(@event.type == pygame.KEYDOWN){
            var keys = pygame.key.get_pressed();
            if(keys[pygame.K_LCTRL]){
                if(@event.key == pygame.K_TAB){
                    graphics.NextUI();
                }
            }else{
                if(@event.key == pygame.K_RETURN){
                    Insert("\n");
                }else if(@event.key == pygame.K_TAB){
                    Insert("    ");
                }else if(@event.key == pygame.K_BACKSPACE){
                    Backspace();
                }else if(@event.key == pygame.K_LEFT){
                    CursorLeft();
                }else if(@event.key == pygame.K_RIGHT){
                    CursorRight();
                }else{
                    Insert(@event.unicode);
                }
            }
        }
    }

    Vector2i GetCursorLocationID(){
        var lineID = 0;
        var numCharacters = 0;
        foreach(var line in lines){
            var charsInLine = len(line);
            if(cursor>=numCharacters && cursor<= numCharacters+charsInLine){
                var cursorIsCharsFromStartOfLine = cursor - numCharacters;
                return (cursorIsCharsFromStartOfLine, lineID);
            }
            numCharacters+=charsInLine+1;
            lineID++;
        }
    }

    void Draw(Graphics graphics, bool active){
        var y = rect[1];
        var x = rect[0];
        var bgcolor = active?(180,180,180):(200,200,200);
        graphics.DrawRect(rect, bgcolor);
        var textColor = (0,20,45);
        var numLines = len(lines);
        foreach(var i in range(scrollY, scrollY+30)){
            if(i>=0 && i<numLines){
                graphics.DrawText((x,y), lines[i], textColor);
                y+=graphics.linesize;
            }
        }

        if(active){
            var cursorYOffset = graphics.fontsize*0.1;
            y = rect[1] - scrollY * graphics.linesize;
            if(len(lines) == 0){
                graphics.DrawRect((x, y, 2, graphics.fontsize), textColor);
            }
            else{
                var locationID = GetCursorLocationID();
                var cursorX = graphics.font.size(lines[locationID[1]][0..locationID[0]])[0] + x;
                var cursorY = graphics.linesize * locationID[1] + y;
                graphics.DrawRect((cursorX, cursorY, 2, graphics.fontsize), textColor);
            }
        }
    }
}

class GameObject{
    string name;
    object[] components;

    GameObject(string _name){
        name = _name;
        components = [];
    }
}

class Button{
    Rect rect;
    string text;
    Action onclick;

    Button(Rect _rect, string _text, Action _onclick){
        rect = _rect;
        text = _text;
        onclick = _onclick;
    }

    void Save(){}

    void OnEvent(Graphics graphics, Event @event){
        if(@event.type == pygame.KEYDOWN){
            if(@event.key == pygame.K_RETURN){
                onclick();
            }else if(@event.key == pygame.K_TAB){
                graphics.NextUI();
            }
        }
    }

    void Draw(Graphics graphics, bool active){
        var bgcolor = active?(180,180,180):(200,200,200);
        graphics.DrawRect(rect, bgcolor);
        var textColor = (0,50,100);
        graphics.DrawText((rect[0],rect[1]), text, textColor);
    }
}

class TreeView{
    Rect rect;
    GameObject[] gameObjects;

    TreeView(Rect _rect){
        gameObjects = [];
        rect = _rect;
    }

    void Save(){}

    void Add(GameObject gameObject){
        gameObjects.append(gameObject);
    }

    void OnEvent(Graphics graphics, Event @event){
        if(@event.type == pygame.KEYDOWN){
            if(@event.key == pygame.K_TAB){
                graphics.NextUI();
            }
        }
    }

    void Draw(Graphics graphics, bool active){
        var bgcolor = active?(180,180,180):(200,200,200);
        graphics.DrawRect(rect, bgcolor);
        graphics.DrawText((rect[0],rect[1]), "TreeView", (0,50,100));
    }
}

class File{
    static string ReadFile(string path){
        var f = open(path, 'r');
        var content = f.read();
        f.close();
        return content;
    }

    static void WriteFile(string path, string content){
        var f = open(path, 'w');
        f.write(content);
        f.close();
    }
}

class Graphics{
    object[] objects;
    object[] uis;
    int activeUI;
    Surface surface;
    Vector2 screenSize;
    Font font;
    float fontsize;
    float linesize;

    Graphics(Vector2 _screenSize, string _font, int _fontsize){
        surface = pygame.display.set_mode((_screenSize[0], _screenSize[1]));
        font = pygame.font.Font(_font, _fontsize);
        fontsize = _fontsize;
        linesize = fontsize * 1.4;
        screenSize = _screenSize;
        uis = [];
        objects = [];
        activeUI = 0;
    }

    void Clear(Color color){
        surface.fill(color);
    }

    void Save(){
        foreach(var u in uis){
            u.Save();
        }
    }

    void DrawText(Vector2 position, string text, Color color){
        var img = font.render(text, true, color);
        surface.blit(img, position);
    }

    void DrawRect(Rect rect, Color color){
        pygame.draw.rect(surface, color, pygame.Rect(rect[0], rect[1], rect[2], rect[3]));
    }

    void AddObject(object obj){
        objects.append(obj);
    }

    void DrawObjects(){
        foreach(var obj in objects){
            obj.Draw();
        }
    }

    void ClearObjects(){
        objects.clear();
    }

    void AddUI(object ui){
        uis.append(ui);
    }

    void OnEvent(Event @event){
        uis[activeUI].OnEvent(this, @event);
    }

    void DrawUI(){
        foreach(var i in range(0, len(uis))){
            uis[i].Draw(this, i==activeUI);
        }
    }

    void NextUI(){
        activeUI++;
        if(activeUI >= len(uis)){
            activeUI = 0;
        }
    }
}

class Program {
    Graphics graphics;

    void Save(){
        graphics.Save();
    }

    void Run(){
        graphics.Save();
        var result = subprocess.run(["compiler\\bin\\Debug\\net8.0\\CSharpToPython.exe", "projects/test/Test.cs--projects/test/Test.py"], 
            shell:true, capture_output:true, text:true).stdout;
        if(result == "ERROR"){
            print(File.ReadFile("projects/test/Test.py"));
        }
        else{
            var globals = {"graphics"=graphics};
            exec(File.ReadFile("projects/test/Test.py"), globals);
        }
    }

    void Start(){
        pygame.init();
        graphics = new Graphics((1280,780), "RedditMono-Medium.ttf", 16);
        graphics.AddUI(new TreeView((20,80,280,680)));
        graphics.AddUI(new CodeEditor((320,80,900,680), File.ReadFile("projects/test/Test.cs")));
        graphics.AddUI(new Button((500,20,100,40), ">", Run));
        graphics.AddUI(new Button((780,20,100,40), "save", Save));

        var running = true;
        pygame.key.set_repeat(500,25);
        while(running){
            foreach(var @event in pygame.@event.@get()){
                if(@event.type == pygame.KEYDOWN && @event.key == pygame.K_ESCAPE){
                    graphics.ClearObjects();
                }else if(@event.type == pygame.QUIT){
                    running = false;
                }else{
                    graphics.OnEvent(@event);
                }
            }
            graphics.Clear((220, 220, 220));
            graphics.DrawUI();
            graphics.DrawObjects();
            pygame.display.flip();
        }
        pygame.quit();
    }

    static void Main(){
        var program = new Program();
        program.Start();
    }
}